contracts/example/SnapshotGate/SnapshotGate.sol

```diff
@@ -0,0 +1,335 @@
+    // token id => owner => total committed
+    mapping(uint256 => mapping(address => uint256)) committed;
+
+    // Modifier to check whether a gate transaction is in progress
+    modifier statusCheck() {
+        txStatus = TransactionStatus.InTransaction;
+        _;
+        txStatus = TransactionStatus.NotInTransaction;
+    }
+
+    // Add safeTransferFrom to IERC20
+    using SafeERC20 for IERC20;
+
+    // Constructor
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        address _protocol,
+        uint256 _sellerId
+    ) ERC721(_name, _symbol) {
+        protocol = _protocol;
+        sellerId = _sellerId;
+        txStatus = TransactionStatus.NotInTransaction;
+    }
+
+    /**
+     * Append a batch of holders to the snapshot
+     *
+     * Reverts if:
+     * - Caller is not contract owner
+     * - Snapshot is frozen
+     *
+     * @param _holders an array of Holder structs
+     */
+    function appendToSnapshot(Holder[] memory _holders) public onlyOwner {
+        require(!snapshotFrozen, "Cannot append to frozen snapshot");
+        uint256 entriesLength = _holders.length;
+
+        // Map all the holders, creating the conditional tokens along the way
+        for (uint256 i = 0; i < entriesLength; i++) {
+            uint256 tokenId = _holders[i].tokenId;
+            uint256 amount = _holders[i].amount;
+            address owner = _holders[i].owner;
+
+            // Store holder's amount of the current snapshot token in mapping
+            snapshot[tokenId][owner] = amount;
+
+            // If corresponding conditional token doesn't yet exist, mint it to custody of this contract
+            if (!_exists(tokenId)) {
+                _mint(address(this), tokenId);
+            }
+        }
+
+        // Notify watchers of state change
+        emit SnapshotAppended(_holders);
+    }
+
+    /**
+     * @notice Freezes the snapshot so that no more holders can be appended
+     *
+     * Reverts if:
+     * - Caller is not contract owner
+     * - Snapshot is already frozen
+     */
+    function freezeSnapshot() external onlyOwner {
+        // Make sure snapshot isn't frozen
+        require(!snapshotFrozen, "Snapshot already frozen");
+
+        // Freeze
+        snapshotFrozen = true;
+
+        // Notify watchers of state change
+        emit SnapshotFrozen();
+    }
+
+    /**
+     * @notice Commit to a gated offer on the Boson Protocol
+     *
+     * Commit to the specified offer on behalf of the buyer,
+     * first checking that the buyer is in the snapshot and
+     * hasn't already used all their available commits.
+     *
+     * Payment must be arranged it the token specified by the
+     * given offer.
+     *
+     * If price is set in the native token, e.g., MATIC on
+     * Polygon, it should be sent to this method in msg.value.
+     *
+     * For all other tokens, advance approval should be done
+     * to allow the protocol to transfer the caller's tokens to
```

Comment:
I think it's more suitable to say "to allow the ~protocol~ **gate** to transfer". 
From buyer's perspective it's imporant that they approve the gate, they don't care about the protocol.